Only transfer one fd at a time

euank opened the pull request:
The previous code would pass all file descriptors as auxiliary data in a
single cmsg. This, it turns out, very easily hits limits on the size of
cmsg aux data. In practice, this happened at around 300 to 500 listening
sockets.

This changes the code to only pass one at a time, adds a test that
failed before, and for good measure switches to a different upstream
library for doing the transfer.

euank committed:
Only transfer one fd at a time

The previous code would pass all file descriptors as auxiliary data in a
single cmsg. This, it turns out, very easily hits limits on the size of
cmsg aux data. In practice, this happened at around 300 to 500 listening
sockets.

This changes the code to only pass one at a time, adds a test that
failed before, and for good measure switches to a different upstream
library for doing the transfer.

inconshreveable reviewed (approved):
lgtm, assuming tests pass. my only concern is that we are now doing O(N) syscalls which could be much slower. i don't think it's worth batching optimizations until/unless it becomes a problem though

euank committed:
Don't close fds in stop

If a user wishes to close file descriptors using the 'Remove' api, they
can.

Unfortunately, the operation of 'Stop' wasn't safe with concurrent
upgrades. Remove is, and callers can use that instead.

euank commented:
Tossed in a commit to avoid the race a test was hitting. There might be a better fix, but as best I could tell it would be quite involved, so I've offloaded a bit more cleanup onto our callers instead.
